packages/discovery/src/discovery/provider/DiscoveryProvider.ts

```diff
@@ -69,40 +69,50 @@ export class DiscoveryProvider {
     topics: Topics,
     fromBlock: number,
     toBlock: number,
+    options: {
+      howManyEvents?: number
+      maxRange?: number
+      filter?: (log: providers.Log) => boolean
+    } = {},
   ): Promise<providers.Log[]> {
     if (fromBlock > toBlock) {
       throw new Error(
         `fromBlock (${fromBlock}) can't be bigger than toBlock (${toBlock})`,
       )
     }
 
-    if (this.getLogsMaxRange === undefined) {
-      return await this.getLogsBatch(address, topics, fromBlock, toBlock)
-    }
-
     // To support efficient caching, we divide the requested blocks range into
-    // sequential boundaries of `maxRange` size, e.g [0,10k-1], [10k, 20k-1], ...
+    // sequential boundaries of `maxRange` size, e.g [10k, 20k-1], [0,10k-1], ...
     // Otherwise ranges would depend on `fromBlock` and even small change to it
     // would make the previous cache useless.
 
-    // Let's start with the deployment block number if it's higher than fromBlock
+    // Let's fetch to the deployment block number if it's higher than fromBlock
     const { blockNumber: deploymentBlockNumber } =
       (await this.getDeploymentInfo(address)) ?? { blockNumber: 0 } // for cases where API to get deployment info is not available
 
-    const maxRange = this.getLogsMaxRange
-    const allLogs: providers.Log[][] = []
-
-    let start = Math.max(fromBlock, deploymentBlockNumber)
+    let allLogs: providers.Log[] = []
+    const maxRange = Math.min(
+      this.getLogsMaxRange ?? Number.MAX_SAFE_INTEGER,
+      options.maxRange ?? Number.MAX_SAFE_INTEGER,
+    )
+    const howManyEvents = options.howManyEvents ?? Number.MAX_SAFE_INTEGER
```

Comment:
I see no reason to use `Number.MAX_SAFE_INTEGER` here. I don't know in what calculations it will be used later. Simply let it be undefined and check for undefined when necessary.